src/share/man/man1/kareboot.1
.\"Generated by db2man.xsl. Don't modify this, modify the source.
.de Sh \" Subsection
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Ip \" List item
.br
.ie \\n(.$>=3 .ne \\$3
.el .ne 3
.IP "\\$1" \\$2
..
.TH "KAREBOOT" 1 "" "" ""
.SH NAME
kareboot \- reboots a node
.SH "SYNOPSIS"
\fBkareboot [\-n|\-\-noreboot] [\-s|\-\-soft] [\-h|\-\-hard] [\-d|\-\-deploy] \-m|\-\-machine hostname [\-e|\-\-environment name] [\-p|\-\-partition partition]\fR
.SH "DESCRIPTION"

.PP
kareboot can execute software or hardware reboots on given nodes\&. It can execute five types of query : no reboot, simple reboot, deploy reboot, environment reboot and partition reboot\&. For all of them, the user can specify one or several hosts\&.

.PP
Here is a brief description of the effects of each reboot type :

.PP
no reboot : setups pxe, grub and tftp stuff for the given nodes without rebooting

.PP
simple reboot : executes a 'normal' (classic) reboot of the given nodes

.PP
deploy reboot : reboots the given nodes on the deployment kernel

.PP
environment reboot : looks for the specified environment on the given nodes and reboots on it\&. The environment should thus already be installed\&.

.PP
partition reboot : reboots the given nodes on the requested partition if an appropriate environment is found

.PP
Consult the example section below for more explicit usage descriptions\&. kareboot needs the appropriate command (softboot, hardboot, deployboot) to be defined in the configuration file in order to be able to execute the requested reboot on the given node ; see configuration file section for further details\&.

.SH "OPTIONS"
\fB\-n, \-\-noreboot\fR
      setups pxe, grub and tftp stuff
      
      \fB\-s, \-\-soft\fR
      requests a software reboot (default reboot type)
      
      \fB\-h, \-\-hard\fR
      requests a hardware reboot

      \fB\-d, \-\-deploy\fR
      specify to reboot on the deployment kernel

      \fB\-m, \-\-machine\fR hostname
      specifies a host\&. This option can be used more than once to specify several hosts\&.
      \fB\-e, \-\-environment\fR
      gives an (already installed) environment to boot\&. When used, only one environment can be specified\&.

      \fB\-p, \-\-partition\fR
      gives a partition to boot on\&. When used, only one partition can be specified\&.

    
.SH "CONFIGURATION FILE"
\fB/etc/kadeploy/deploy_cmd\&.conf\fRCan contain four entry types per node whose three are relative to the kareboot command\&. Those three entries describe the reboot commands for the soft, hard and deploy reboot type\&. Here is the description of the expected syntax : 
      
      host_name reboot_type command
      
      where reboot_type can be either softboot, hardboot or deployboot\&. 
      
      Example for idpot1 of idpot cluster :

      
.PP
idpot1 softboot ssh root@idpot1 \-o ConnectTimeout=2 /sbin/reboot

.PP
idpot1 hardboot ssh localuser@ldap\-idpot ~/commande_modules/logiciel/dio_reset 0x81 0

.PP
idpot1 deployboot \&.\&./boot/setup_pxe\&.pl 192\&.168\&.10\&.1:label_deploy_tg3

      
      In this example, 

      
.PP
soft reboot is actually a simple "reboot" command done via a ssh connection as root

.PP
hard reboot is a hard signal sent via a ssh connection on a precise server as a special user

.PP
deploy boot is a call to the (deployment system) setup_pxe\&.pl perl script with appropriate arguments

.SH "EXAMPLE"

.PP
\fINo reboot : \fR

.PP
\fBkareboot \-n \-m node1 \-e debian \fR setups pxe, grub and tftp stuff for rebooting node1 on debian environment (if already installed)\&.

.PP
\fBkareboot \-n \-m node7 \-p hda5 \fR setups pxe, grub and tftp stuff for rebooting node7 on partition hda5 (if an environment is found)\&.

.PP
\fISimple reboot : \fR

.PP
\fBkareboot \-s \-m node1\fR executes a software reboot of node1\&.

.PP
\fBkareboot \-h \-m node7 \-m node6\fR executes a hardware reboot of nodes 7 and 6\&.

.PP
\fBkareboot \-m node7\fR executes a software reboot of node7\&.

.PP
\fIDeploy reboot : \fR 

.PP
\fBkareboot \-d \-m node2\fR (soft) reboots on deployment kernel\&.

.PP
\fBkareboot \-d \-h \-m node2\fR (hard) reboots on deployment kernel\&.

.PP
\fIEnvironment reboot : \fR 

.PP
\fBkareboot \-e debian \-m node3\fR looks for environment debian on node3 and (soft) reboots on it if it exists\&.

.PP
\fBkareboot \-h \-e debian \-m node3\fR looks for environment debian on node3 and (hard) reboots on it if it exists\&.

.PP
\fIPartition reboot : \fR

.PP
\fBkareboot \-s \-p hda6 \-m node4\fR checks the environment installed on partition hda6 of node4 and (soft) reboots on it if appropriate\&.

.SH "COPYRIGHTS"

.PP
Copyright 2004 Laboratoire Id\-Imag (http://www\-id\&.imag\&.fr)\&. This software is licensed under the GNU Library General Public License\&. There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE\&.



src/share/perl/5.8/libkadeploy2/kareboot.pm
package libkadeploy2::kareboot;

use File::Copy;
use Getopt::Long;
use libkadeploy2::deployconf;
use libkadeploy2::deploy_iolib;
use libkadeploy2::command;
use libkadeploy2::message;
use libkadeploy2::nodelist;
use libkadeploy2::cmdline;
use libkadeploy2::karights;
use libkadeploy2::sudo;
use strict;
use warnings;

sub execcmd($$$);
sub check_options;
sub check_right($);


my $conf=libkadeploy2::deployconf::new();
if (! $conf->loadcheck()) { exit 1; }

my $sudo_user=libkadeploy2::sudo::get_sudo_user();
if (! $sudo_user) { $sudo_user=libkadeploy2::sudo::get_user(); }

my $ref_node_list;
my @node_list;
my $node_file;
my $node;
my $env;
my $soft;
my $hard;
my $deploy;
my $noreboot;
my $device;
my $verbose=0;
my $refnodelist;
my $command;
my $cmd;
my $nodeshell;
my $nodelist;
my $help;
my $ok=1;
my $message=libkadeploy2::message::new();
my $kadeploydir=$conf->get("kadeploy2_directory");
my $righttocheck="REBOOT";

sub run()
{
    if (! check_options()) { return 1; }

    if (! check_right($righttocheck)) { $message->message(2,"$sudo_user not allowed to $righttocheck ".$nodelist->get_str()); exit 1; }
    
    
    $ref_node_list=$nodelist->get_nodes();
    @node_list=@$ref_node_list;
    
    foreach $node (@node_list)
    {
	if ($soft)
	{
	    if (! execcmd($node,"softboot",$verbose))
	    { $ok=0; }
	}    
	if ($hard)
	{
	    if (! execcmd($node,"hardboot",$verbose))
	    { $ok=0; }
	}
	if ((! $soft ) &&
	    (! $hard))
	{
	    if (! execcmd($node,"softboot",$verbose))
	    {
		if (! execcmd($node,"hardboot",$verbose))
		{		
		    $ok=0;
		}
	    }
	}
    }
    if (! $ok) { return 1; }
    else       { return 0; }
}

################################################################################

sub get_options_cmdline()
{
    GetOptions(
	       'm=s'           => \@node_list,
	       'machine=s'     => \@node_list,
	       'f=s'           => \$node_file,

	       'e=s'           => \$env,
	       'environment=s' => \$env,

	       's'             => \$soft,
	       'soft'          => \$soft,

	       'h'             => \$hard,
	       'hard'          => \$hard,

	       'd'             => \$deploy,
	       'deploy'        => \$deploy,

	       'n'             => \$noreboot,
	       'noreboot'      => \$noreboot,

	       'p=s'           => \$device,
	       'partition=s'   => \$device,

	       'verbose'       => \$verbose,
	       'v'             => \$verbose,

	       'help!'         => \$help,
	       'h!'            => \$help,
	       );

}

sub check_options()
{

    if ($help) { $message->kareboot_help(); exit 0; }

    if (@node_list)     { $nodelist=libkadeploy2::nodelist::new(); $nodelist->loadlist(\@node_list);  }
    else { $message->missing_node_cmdline(2); return 0; }


    if (!scalar($nodelist))
    {
	$message->kareboot_help();
	return 0;
    }
    
    if ($soft && $hard)
    {
	print "ERROR : soft and hard reboot options are exclusive\n";
	print "ERROR : please select only one of them at once\n";
	return 0;
    }
    
    if (($noreboot && $soft) || ($noreboot && $hard))
    {
	print "ERROR : no reboot and soft or hard reboot should be exclusive\n";
	print "ERROR : please select only one of them at once\n";
	return 0;
    }
    
    if (($deploy && $env) || ($deploy && $device) || ($env && $device))
    {
	print "ERROR : please select EITHER a deployment reboot OR a reboot on a partition OR a reboot on an environment\n";
	return 0;
    }

    return 1;
}



sub check_right($)
{
    my $righttocheck=shift;
    return libkadeploy2::karights::check_rights($nodelist,$righttocheck);
}



sub execcmd($$$)
{
    my $node=shift;
    my $cmdnode=shift;
    my $verbose=shift;
    my $refnodelist;
    my @node_list;
    my $ok=1;
    
    $refnodelist=$nodelist->get_nodes();
    @node_list=@$refnodelist;


    $cmd=$conf->getpath_cmd("kaexec")." --confcommand -m ".$node->get_name()." -c $cmdnode";
    $command=libkadeploy2::command::new($cmd,30,$verbose);
    $command->exec();
    if (! $command->get_status())
    { 
	$message->message(0,$sudo_user." ".$cmdnode." ".$node->get_name()." fail");
	$ok=0; 
    }
    else
    {
	$message->message(0,$sudo_user." ".$cmdnode." ".$node->get_name()." success");
    }    
    return $ok;
}



1;
